### DISCUSSION TITLE
Help parsing Google's SSML

### DISCUSSION THREAD

**Discussion by gad2103** [pdvrieze/xmlutil#74] [Q&A] [answered]
Hello. First, thanks for this awesome library! I'm getting stuck on something. The simplest version of it is something I can't wrap my head around. my understanding of XML is pretty shallow, so i thought after spending 4 hours trying to figure this out i'd ask the experts. 

here's what i've done so far:
- read the tests in this repo
- updated to latest lib release
- simplified to a simple toy example

Here is the simplest picture of what i'm trying to do. Say i have some xml like:
```
val expected = """
            <speak data-ui-title="hello">
                <par>
                    <media xml:id="prompt">some text here<audio src="234" /></media>
                    <media xml:id="speak" begin="prompt.end-1s">
                    <audio soundLevel="-15dB" src="https://soundsrc.com"/>
                    </media>
                </par>
            </speak>"""
```

I am parsing this using
```
val xmlSerializer = XML(SpeakTag.module) {
    indentString = "    "
    policy = DefaultXmlSerializationPolicy(
        pedantic = false,
        autoPolymorphic = true
    )
}

fun decodeXmlFromString(ssml: String): SpeakTag {
    return xmlSerializer.decodeFromString<SpeakTag>(ssml)
}

fun SSML.decodeXmlFromValue(): SpeakTag {
    return xmlSerializer.decodeFromString<SpeakTag>(value)
}

interface MediaChildren
interface SpeakChildren

/**
 * currently this is inflexible and would break on certain formats of ssml
 * a full wrapper of google's ssml spec would need to be written to handle
 * both arbitrary nestings and arrangements + extra tags
 */
@Serializable
@SerialName("speak")
data class SpeakTag(
    @XmlValue(true)
    val content: String? = null,
    @SerialName("data-intent-id")
    val id: Double? = null,
    @SerialName("data-ui-title")
    val responseTitle: String? = null,
    @SerialName("data-ui-intent")
    val intentName: String? = null,
    @SerialName("data-ui-hide")
    val hideResponse: Boolean? = null,
    @SerialName("data-ui-hide-intent")
    val hideIntent: Boolean? = null,
    @SerialName("data-condition")
    val dataCondition: String? = null,
    @XmlValue(true)
    val data: @Polymorphic SpeakChildren? = null,
): MediaChildren {
    companion object {
        val module = SerializersModule {
            polymorphic(SpeakChildren::class) {
                subclass(ParallelTag::class)
            }
            polymorphic(MediaChildren::class) {
                subclass(AudioTag::class)
                subclass(SpeakTag::class)
            }
        }
    }

    override fun toString(): String = xmlSerializer.encodeToString(this)
}

@Serializable
@SerialName("par")
data class ParallelTag(
    val data: List<Media> = listOf()
): SpeakChildren

@Serializable
@SerialName("audio")
data class AudioTag(
    val soundLevel: String? = null,
    val src: String? = null
): MediaChildren

@Serializable
@SerialName("media")
data class Media(
    @XmlSerialName("id", "", "xml")
    val id: String? = null,
    @SerialName("begin")
    val startId: String? = null,
    val fadeInDur: String? = null,
    @XmlValue(true)
    val content: String? = null,
    @XmlValue(true)
    val data: @Polymorphic MediaChildren? = null
)
```

the tricky part is the `<media xml:id="something"><audio /></media>`

the serializer **decodes XML strings just fine**  meaning it will parse the xml string to data classes without an issue. however, it does not **encodeToString** as I would expect. When i encode it to a string I get the following:

```
            <speak data-ui-title="hello">
                <par>
                    <media id="prompt">some text here<audio src="234" /></media>
                    <media id="speak" begin="prompt.end-1s">
                    <audio soundLevel="-15dB" src="https://soundsrc.com"/>
                    </media>
                </par>
            </speak>
```

This isn't valid for the use case i have. The `xml:id` needs to be in the encoded string for the SSML to parse correctly.
I've tried a few other things. If i do the following change, everything **encodes to string** just fine but **will not decodeFromString** properly
```
@Serializable
@SerialName("media")
data class Media(
    @SerialName("xml:id")
    val id: String? = null,
```

so it seems i can only get *ONE* direction to work at a time. Please let me know what I'm doing wrong here. Thanks again!!!!

***UPDATE***
After updating the library from `0.82.1` to `0.83.0` it looks like what i just said ☝️  is no longer accurate. Now When i try to decode with 
```
@Serializable
@SerialName("media")
data class Media(
    @SerialName("xml:id")
    val id: String? = null,
```
i get 
```
Could not find a field for name {http://www.w3.org/XML/1998/namespace}id
  candidates: xml:id, begin, fadeInDur, content, speak, audio at position Line number = 3
Column number = 32
System Id = null
Public Id = null
Location Uri= null
CharacterOffset = 72
```
in fact, every attempt to decode throws that error no matter what i use for the annotation 😭

**Comment by pdvrieze** [comment, marked as answer]
First of all (for the updated bit), you should probably use `@XmlSerialName` that allows specifying names. The format uses namespaces, "assumes" that there are no special characters in the serial name. You can create your own policy to try to interpret it differently (there is no single way to transform serial names into qnames).

For xml:id, you will need to specify the proper namespace. When serializing, the prefix is used only on an advisory level (because prefixes are not semantically relevant, and multiple prefixes for the same namespace create unneeded noise). The value you want is:
` @XmlSerialName("id", "http://www.w3.org/XML/1998/namespace", "xml")`.

**note** while the prefix xml is forced to that namespace (as the standards specify), the decoder does not do this.

**Reply by gad2103** [reply]
thanks for the quick reply. this is a dumb question, but how does that namespace work? if i put in a random string would it still parse correctly? or is that URL actually used in the parsing?

**Reply by pdvrieze** [reply]
Namespaces allows you to use a combination of url+local name to determine what the name means. A prefix is used to bind a url to a tag as it is a lot shorter. The prefix is local to your document and it doesn't matter what value (with legal characters) it has. So the namespace cannot be random, the prefix can be (not starting with "xml" for non-standard namespaces, within legal characters - leter-ish and no ":" ). In this particular case the id attribute used is the one from the standards xml:id
